app/services/fixity_checker.rb

```diff
@@ -0,0 +1,117 @@
+  end
+
+  #FixityChecker.new(asset).check
+  def check(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return nil if @asset.nil?
+    return nil if @asset.file.nil?
+    the_expected_checksum = expected_checksum
+    # Note: @asset.file.sha512 can be nil on e.g.
+    # items imported with DISABLE_BYTESTREAM_IMPORT=true
+    # For now, we are ignoring these.
+    return nil if the_expected_checksum.nil?
+
+    new_check = FixityCheck.new(asset: @asset)
+    new_check.expected_result = the_expected_checksum
+    new_check.checked_uri = @asset.file.url
+    new_check.actual_result = actual_checksum
+    new_check.passed= (the_expected_checksum==actual_checksum)
+    new_check.save!
+  end
+
+  #FixityChecker.new(asset).prune_checks
+  # After pruning, for each asset - URI combination, you will be left with:
+  # up to NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP recent passed checks,
+  # plus all the failed checks,
+  # plus one initial check (to remember the file that was intially uploaded.)
+  def prune_checks(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return if @asset.file.nil?
+    return if @asset.file.url.nil?
+    checks_its_ok_to_delete.map(&:destroy)
+  end
+
+  private
+
+  # Returns an array of checks that we don't want to keep.
+  # The current rules are:
+  # Never throw out FAILED checks
+  # Never throw out the earliest PASSED check
+  # Always keep N recent PASSED checks
+  def checks_its_ok_to_delete
+    # Throw all the passed checks INTO the trash.
+    checks = FixityCheck.checks_for(@asset, @asset.file.url)
+    return [] if checks.empty?
+    trash = checks.select { | ch | ch.passed? }
+    # But then, pop the earliest passed check back OUT of the trash.
+    earliest_passed_check = trash.pop
+    # Finally, shift the most recent N passed checks back OUT of the trash.
+    recent_passed_checks = trash.shift(NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP)
+    return trash
+  end
+
+  def expected_checksum
+    @asset.file.sha512
+  end
+
+
+  # Sifts all our  assets by some integer
+  # between one and CHECK_CYCLE_LENGTH.
+  # Allows us to convieniently check only a subset of
+  # the assets at a time, but be sure everything eventually
+  # gets checked.
+  # Pass in 0, and everything will go through the sieve.
+  def sift(sieve_integer)
+    return true if sieve_integer == 0
+    check_sieve(sieve_integer)
+    @asset.friendlier_id.bytes.sum % CHECK_CYCLE_LENGTH == sieve_integer
+  end
+
+  # Check this is a positive int between 1 and the cycle length.
+  def check_sieve(sieve_integer)
+    if !(sieve_integer.is_a? Integer)        ||
+      sieve_integer < 1                      ||
+      sieve_integer > CHECK_CYCLE_LENGTH
+      raise ArgumentError.new(
+        "Expected a positive int between  1 and #{CHECK_CYCLE_LENGTH}. Got #{sieve_integer}"
+      )
+    end
+  end
+
+  def actual_checksum
+    sha512 = Digest::SHA512.new
+    @asset.file.open(rewindable:false) do |io|
+      if io.is_a? File
```

Comment:
Curious in what cases you were getting a `File` here?  Does `each_chunk` not work on a `File`? I see it does not! Annoying that we need two different paths here depending on what we get from `file.open`, I'm going to bring it up with janko from shrine.

Reply:
Okay, here's how shrine actually does calculation:  https://github.com/shrinerb/shrine/blob/master/lib/shrine/plugins/signature.rb#L101-L106

So, if we do it shrine's way a) I trust that janko knows what he's doing it and it's efficient, and b) it should work with File or Down::ChunkedIO without requiring a separate path.

It might look like this:

```ruby
@asset.file.open(rewindable:false) do |io|
   digest = Digest::SHA512.new
   digest.update(io.read(16*1024, buffer ||= String.new)) until io.eof?
   digest.hexdigest
end
```

(Tested in a console, that worked. Changed shrine's `.digest` to `.hexdigest` to get the hexadecimal one we want instead of binary bytes.)


We could leave a comment linking to shrine's code saying we copied that. On the other hand, this is weird code, maybe your simpler version is better?

Reply:
Used the shrine recipe-- looks great to me.